### ISSUE TITLE
Training questions

### ISSUE THREAD

**Issue by aldll** [lyst/lightfm#317]
Hi,
I have two questions,
1. I try to build a recommender 'online'. If my input data(interaction, user features) has some new actions, and I want to update the model has trained before. Should I retrain a new model? 
or I can just use method `fit_partial` like this
`model.fit_partial(updated_interaction, updated_user_features, item_features)`
to update my model?

2. I observed the model training with `num_thread=4` on linux server but it seems only use one CPU and the other three use 3% - 5.5% only. 
I saw that LightFM uses OpenMP but I'm not run it on OSX or Windows. 
Also I reinstall LightFM and not see any warning like `LightFM was compiled without OpenMP support.`, so OpenMP should be enable? How to run my training to use four CPUs with 100%?
Is there something wrong? How can I check it?

Sorry for long description about my questions, thank you!

**Comment by maciejkula** [comment]
1. You can use `fit_partial` as you suggest.
2. How do you measure this? You will only see full utilization in `top/htop` if you run a long-running fitting process: 10 epochs on Movielens 100K runs too quickly for `top` to show the utilization. Can you run it with something like 1000 epochs and see?

**Comment by aldll** [comment]
1. OK I see! Thank you!
2. I have tried epochs 100 300, now I also try 1000, but the CPU utilization still the same(only one is full utilized), so it makes me confused.

**Comment by maciejkula** [comment]
Can you post the code you are using? Did you install LightFM from conda or from pip?

**Comment by aldll** [comment]
I install LightFM by pip, `pip3 install lightfm`
my code:
```
model = LightFM(loss='warp', no_components=100, random_state=seed)
model.fit_partial(interaction,
                      user_features=user_features,
                      item_features=item_features,
                      epochs=100,
                      num_threads=4)
```

**Comment by maciejkula** [comment]
I'm afraid I cannot reproduce your problem. Can you run the following and post the output?
```
echo |cpp -fopenmp -dM |grep -i open
```

**Comment by aldll** [comment]
OK
the output
`#define _OPENMP 201107`

**Comment by maciejkula** [comment]
Do you use a lot of regularization?

Other than this, nothing comes to mind. Can you post a gist of code that reproduces the problem on the built-in Movielens dataset?

**Comment by maciejkula** [comment]
That looks fine. Are you using a lot of regularisation?

**Comment by aldll** [comment]
Is regularization some parameters in model?
I used the code above to build the model, and the other settings are default, so I think I didn't use regularization?

**Comment by maciejkula** [comment]
It looks like you didn't.

I'm afraid I can't help more unless I can reproduce.

**Comment by aldll** [comment]
I have trying to figure out the problem. Sorry for replying late,
I want to make sure that:
Is the regularzation means normalization?
Because I used normalization in my item features twice,
At first stage, I build a tf-idf matrix and normalizing it to unit length.
Then I use L1 normalize to rescale tf-idf matrix so that each row's sum equals to 1.

I tried to confirm the assumption above, so I use other item features with all value is 1 to train model, it can run with 4 CPUs.
Is the item features can't contain too many small values for algorithm of SGD to compute?
Could you please tell me description about the question? I want to know it in more details. And what is the problem about using a lot of regularization? Thank you for reading my long questions! I appreciate it!

**Comment by maciejkula** [comment]
By regularization I mean the L2 regularization imposed by the `item/user_alpha` hyperparameter.

The algorithm forces full regularization passes whenever the accumulated regularization is quite large. It's possible you're experiencing this, although it shouldn't be dependent on the feature sizes.

At this stage you _must_ give me a short example on synthetic data that reproduces this. Otherwise we're not going to get anywhere.

**Comment by aldll** [comment]
OK, this is the data: https://github.com/aldll/t
Since the data is too large, I only upload 3000 items.

You said "The algorithm forces full regularization passes whenever the accumulated regularization is quite large."
Sorry I didn't understand this meaning. Could you explain more about it? Many thanks!

**Comment by maciejkula** [comment]
There are 8 interactions in your data. You are not going to get parallelism on data that small. I can confirm that increasing the number of interactions to 100000 via
```
    n = 100000
    interaction = sp.coo_matrix((np.ones(n), (np.arange(n), np.ones(n))))
```
makes the parallelism kick in.

**Comment by aldll** [comment]
Got it!
May I ask for more details in algorithm you mention before? 
Also, the recommend result with different num_threads is different, it there any setting wrong?
Or the reason is the algorithm?

**Comment by maciejkula** [comment]
When using parallelism the results are not deterministic, and different than just using a single thread. Have a look at [the Hogwild paper](https://people.eecs.berkeley.edu/~brecht/papers/hogwildTR.pdf) for an idea about the reasons.

**Comment by aldll** [comment]
I see. I will read the paper!
Thank you for your answer and provide great information!
I will close this issue later.
